### PR TITLE
Add feature to delete RDS snapshots

### PR DESCRIPTION
**Pull request by alexkayabula**
### Description
Adding support for filtering and deleting RDS snapshots by tag, date, name.

### Related Issue
Closes: #120  

### Tasks

- [x] Get all RDS snapshots
- [x] Nuke/Delete collected RDS snapshots
- [x] Enable CLI configuration
- [x] Add `rdssnapshots` resource to list of resources
- [x] Get all Aurora RDS Cluster snapshots
- [x] Nuke/Delete all collected Aurora RDS Cluster snapshots
- [x] Add RDS Instance snapshots tests
- [x] Add Aurora RDS Cluster snapshots tests
- [x] Implement filtering by names_regex
- [x] Implement filtering by tags_regex

### CODE REVIEW

**Comment by alexkayabula** [comment]
@brikis98 @yorinasub17 Please review and check this PR and see if meets expectation. I am also kindly request that you don't snooze it like [125]( https://github.com/gruntwork-io/cloud-nuke/pull/125).

**Comment by saurabh-hirani** [comment]
> @brikis98 @yorinasub17 Please review and check this PR and see if meets expectation. I also kindly request that you don't snooze it like [125](https://github.com/gruntwork-io/cloud-nuke/pull/125).

I think you clearly meant this PR - https://github.com/gruntwork-io/cloud-nuke/pull/114 which has the aforementioned comment on delayed review. 

@alexkayabula - I have also put in personal time and weekends to contribute to open source. While I understand the review delay from the project maintainers as they might have other priorities, seeing someone make an example of that delay is plain hurtful and derides my contribution. Not cool and totally uncalled for.

**Comment by saurabh-hirani** [comment]
Attaching screenshot of the edited comment - as I suspected that instead of engaging to resolve the point I raised  - my concern would be totally ignored and the comment deleted/edited. 

Screenshot 2020-07-14 at 4 03 56 PM

**Comment by brikis98** [comment]
Hey folks, we're doing the best we can. I understand you're frustrated at the delays, but I kindly request that you don't take it out on us. Like everyone else, we're doing our best to balance open source maintenance, with running and growing a small business, and all this amidst COVID-19, which, for many people, has affected available hours (e.g., some of the maintainers have families who are now at home all day and, I hope you agree, take higher priority than responding to issues/PRs).

Not sure why you felt the need for the screenshot. @alexkayabula edited his own comment (see the comment history). We've never edited or deleted other people's comments and, in general, have a [strong dedication to transparency and honesty](https://blog.gruntwork.io/how-we-built-a-distributed-self-funded-family-friendly-profitable-startup-93635feb5ace), so it's not clear where this comes from. 

We greatly appreciate your contributions and we'll incorporate them as soon as we can. In the meantime, grab a 🍺 or ☕, relax, and try to keep the conversation positive 👍

**Comment by alexkayabula** [comment]
@saurabh-hirani I honestly understand the frustration but I suspect that @brikis98 and @yorinasub17 could be tigthly engaged at the moment. @brikis98 usually promptly reviews other open source projects like [terragrunt](https://github.com/gruntwork-io/terragrunt) that I have been contributing to. Lets try to keep the conversation positive and fun.

**Comment by saurabh-hirani** [comment]
I am all for positivity - having the opportunity to contribute to open source is a privilege and to engage with the best minds in the industry even more so. I look up to the work that Gruntwork does and can only aspire to be emulate their level of work. 

Also the point isn't about delayed reviews - I am fine with a delayed review as I called in my original comment. I go through Gruntwork's monthly newsletter and it's amazing how they contribute so much to open source while creating production grade Terraform modules.  I just felt calling out my contribution to push a case (as in @alexkayabula original comment) was not cool. The original comment was edited and I meant that Alex edited it instead of addressing the point I raised and I didn't mean anything else. I mentioned that as it would be easier to just keep the original comment as is and engage. But eitherways, I would want to bury the hatchet and move on. Thanks for addressing this @brikis98 and sorry for digressing this thread from a tech review. Cheers! 🍻

**Comment by brikis98** [comment]
Understood. Thanks everyone! 👍  🍻

File: aws/rds_cluster_snapshots_test.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		DBClusterSnapshotIdentifier: &snapshotName,
+	})
+
+	// Generate a tag
+	var tags = []*rds.Tag{
+		{
+			Key:   awsgo.String("snapshot-tag1-key"),
+			Value: awsgo.String("snapshot-tag1-value"),
+		},
+	}
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:100]: NIT: The method description doesn't mention anything about tagging the snapshot. Perhaps update the description? And/or take in the tag key/value as params?

File: aws/rds_cluster_snapshots_test.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+	// Tag a DB Cluster snapshot resource
+	for _, database := range result.DBClusterSnapshots {
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:103]: Oh, wait, this tags _all_ the snapshots for the DB... Is that intentional? If it is, it should probably be extracted into a separate method that makes it clear that'll happen.

File: aws/rds_cluster_snapshots_test.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+
+	// Create a test DB Cluster and use the dbClusterIdentifier to generate snapshots
+	clusterName := "cloud-nuke-test-cluster" + util.UniqueID()
+	createTestDBCluster(t, session, clusterName)
+	excludeAfter := time.Now().Add(1 * time.Hour)
+	clusterNames, err := getAllRdsClusters(session, excludeAfter)
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:129]: Check `err`!

File: aws/rds_cluster_snapshots_test.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	dbClusterIdentifier := awsgo.StringValueSlice(clusterNames)[0]
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:130]: Check `clusterNames` is not empty!

File: aws/rds_cluster_snapshots_test.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+	require.NoError(t, errors.WithStackTrace(err))
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+
+	// Create a test DB Cluster and use the dbClusterIdentifier to generate snapshots
+	clusterName := "cloud-nuke-test-cluster" + util.UniqueID()
+	createTestDBCluster(t, session, clusterName)
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:127]: Need to `defer` cleaning up this cluster

File: aws/rds_cluster_snapshots_test.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		svc.AddTagsToResource(&rds.AddTagsToResourceInput{
+			ResourceName: awsgo.String(*arn),
+			Tags:         tags,
+		})
+	}
+
+}
+
+// TestFilterRDSClusterSnapshot tests for filtering cluster snapshots basing on matching include and exclude rules in the config file.
+func TestFilterRDSClusterSnapshot(t *testing.T) {
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_test.go:114]: How long does this test take to run?

File: aws/rds_cluster_snapshots_types.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+	return nil
+}
+
+type RdsClusterSnapshotDeleteError struct {
+	name string
+}
+
+func (e RdsClusterSnapshotDeleteError) Error() string {
+	return "RDS DB Cluster Snapshot:" + e.name + "was not deleted"
```
Comment (brikis98) [inline aws/rds_cluster_snapshots_types.go:42]: Spaces around the name are missing. It's more idiomatic to use `fmt.Sprintf` for this sort of thing, which helps avoid this sort of error:

```suggestion
	return fmt.Sprintf("RDS DB Cluster Snapshot %s was not deleted", e.name)
```

Please use the same idea to the other errors below too.

File: aws/rds_snapshot.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// Get all DB Instance snapshots
+func getAllRdsSnapshots(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	svc := rds.New(session)
+
+	result, err := svc.DescribeDBSnapshots(&rds.DescribeDBSnapshotsInput{})
```
Comment (brikis98) [inline aws/rds_snapshot.go:19]: Same question about pagination

File: aws/rds_snapshot.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+
+		// Automated DB snapshots can only be deleted by deleting the DB instance or
+		// changing the backup retention period for the DB instance to 0.
+		// This edge case can't be handled since all DB instance related automated snapshots will be deleted.
+		// Refer to https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_DeleteSnapshot.html
+		if *snapshot.SnapshotType != "automated" {
```
Comment (brikis98) [inline aws/rds_snapshot.go:42]: Same comment on extracting the nested code into a separate function

File: aws/rds_snapshot.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+	} else {
+		// Otherwise
+		shouldInclude = true
+	}
+
+	return shouldInclude
+}
+
+// Include filtered DB Instance snapshot
+func includeSnapshotByNamesREList(snapshotName string, reList []*regexp.Regexp) bool {
```
Comment (brikis98) [inline aws/rds_snapshot.go:84]: Same comment on a single function for handling these checks

File: aws/rds_snapshot_test.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+// createTestDBInstance generates a test DBInstance since snapshots can only be created
+// from an existing DB Instance
+func createTestDBInstance(t *testing.T, session *session.Session, name string) {
+	svc := rds.New(session)
+	input := &rds.CreateDBInstanceInput{
+		AllocatedStorage:     awsgo.Int64(5),
+		DBInstanceClass:      awsgo.String("db.m5.large"),
```
Comment (brikis98) [inline aws/rds_snapshot_test.go:25]: Is this the smallest instance type we can use?

File: aws/rds_snapshot_test.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+	_, err := svc.CreateDBSnapshot(input)
+	require.NoError(t, errors.WithStackTrace(err))
+
+	svc.WaitUntilDBSnapshotAvailable(&rds.DescribeDBSnapshotsInput{
+		DBInstanceIdentifier: &instanceName,
+		DBSnapshotIdentifier: &snapshotName,
+	})
+
+	result, err := svc.DescribeDBSnapshots(&rds.DescribeDBSnapshotsInput{
```
Comment (brikis98) [inline aws/rds_snapshot_test.go:56]: Check `err`! 

BTW, many editors with good Golang support can warn you about ignored errors automatically.

File: aws/rds_snapshot_test.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	// Generate a tag
+	var tags = []*rds.Tag{
+		{
+			Key:   awsgo.String("snapshot-tag1-key"),
+			Value: awsgo.String("snapshot-tag1-value"),
+		},
+	}
+
+	// Tag the DB snapshot resource
+	for _, database := range result.DBSnapshots {
```
Comment (brikis98) [inline aws/rds_snapshot_test.go:69]: Same comment about this method both creating a snapshot and tagging all existing snapshots; should probably be multiple methods.

File: aws/rds_snapshot_test.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	require.NoError(t, errors.WithStackTrace(err))
+
+	session, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region)},
+	)
+
+	// Create a test DB Instance and use the dbInstanceIdentifier to generate snapshots
+	instanceName := "cloud-nuke-test-instance-" + util.UniqueID()
+	excludeAfter := time.Now().Add(1 * time.Hour)
+	createTestDBInstance(t, session, instanceName)
```
Comment (brikis98) [inline aws/rds_snapshot_test.go:93]: Need to `defer` cleaning up this DB

File: aws/rds_snapshot_types.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+type DBSnapshots struct {
+	SnapshotNames []string
+}
+
+// Name of the AWS resource
+func (snapshot DBSnapshots) ResourceName() string {
+	return "rdssnapshots"
```
Comment (brikis98) [inline aws/rds_snapshot_types.go:15]: Name should be different!

File: aws/rds_cluster_snapshots.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+
+		// Automated DB snapshots can only be deleted by deleting the DB instance or
+		// changing the backup retention period for the DB instance to 0.
+		// This edge case can't be handled since all DB instance related automated snapshots will be deleted.
+		// Refer to https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_DeleteSnapshot.html
+		if *snapshot.SnapshotType != "automated" {
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Fixed

File: aws/rds_cluster_snapshots_test.go [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+		DBClusterIdentifier: name,
+	}
+
+	for i := 0; i < 90; i++ {
+		_, err := svc.DescribeDBClusters(input)
+		if err != nil {
+			return err
+		}
+
+		time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:34]: Fixed

File: aws/rds_cluster_snapshots_test.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+		DBClusterIdentifier:         clusterName,
+		DBClusterSnapshotIdentifier: snapshotName,
+	}
+	for i := 0; i < 90; i++ {
+		_, err := svc.DescribeDBClusterSnapshots(input)
+		if err != nil {
+			return err
+		}
+
+		time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:54]: Fixed

File: aws/rds_cluster_snapshots_test.go [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+	input := &rds.CreateDBClusterInput{
+		DBClusterIdentifier: awsgo.String(name),
+		Engine:              awsgo.String("aurora-mysql"),
+		MasterUsername:      awsgo.String("gruntwork"),
+		MasterUserPassword:  awsgo.String("password"),
+	}
+
+	_, err := svc.CreateDBCluster(input)
+	require.NoError(t, err)
+	waitUntilRdsClusterAvailable(svc, &name)
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:76]: Fixed

File: aws/rds_snapshot_test.go (hunk 14 above)
Comment (alexkayabula) [reply aws/rds_snapshot_test.go:97]: Fixed

File: aws/rds_snapshot_test.go (hunk 12 above)
Comment (alexkayabula) [reply aws/rds_snapshot_test.go:56]: Fixed

File: aws/rds_cluster_snapshots_test.go [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:77]: Fixed

File: aws/rds_cluster_snapshots_test.go (hunk 5 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:132]: Fixed

File: aws/rds_cluster_snapshots_test.go (hunk 3 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:134]: Fixed

File: aws/rds_cluster_snapshots_test.go (hunk 4 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:142]: Fixed

File: aws/rds_cluster_snapshots_test.go (hunk 2 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go]: Fixed

File: aws/rds_cluster_snapshots_test.go (hunk 1 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go]: Fixed

File: aws/rds_cluster_snapshots.go [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+	} else {
+		// Otherwise
+		shouldInclude = true
+	}
+
+	return shouldInclude
+}
+
+// Include filtered DB Cluster if tag matches
+func includeClusterSnapshotByTagsREList(tagName string, reList []*regexp.Regexp) bool {
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go:147]: Thank you for pointing that out. Let me refactor.

File: aws/rds_cluster_snapshots.go [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+			}
+		}
+	}
+
+	return snapshots, nil
+}
+
+// Filter DB Cluster snapshot by names_regex in config file
+func shouldIncludeClusterSnapshotByName(snapshotName string, includeNamesREList []*regexp.Regexp, excludeNamesREList []*regexp.Regexp) bool {
+	shouldInclude := false
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Fixed.

File: aws/rds_cluster_snapshots.go [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+		if re.MatchString(snapshotName) {
+			return false
+		}
+	}
+	return true
+}
+
+// Filter DB Cluster snapshot by tags_regex in config file
+func shouldIncludeClusterSnapshotByTag(tagName string, includeTagNamesREList []*regexp.Regexp, excludeTagNamesREList []*regexp.Regexp) bool {
+	shouldInclude := false
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Fixed

[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Fixed

File: aws/rds_cluster_snapshots.go [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	return RdsClusterSnapshotDeleteError{name: *input.DBClusterSnapshotIdentifier}
+}
+
+// Get all DB Cluster snapshots
+func getAllRdsClusterSnapshots(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	svc := rds.New(session)
+
+	result, err := svc.DescribeDBClusterSnapshots(&rds.DescribeDBClusterSnapshotsInput{})
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Yes it is. Thank you for pointing that out. Fixed.

File: aws/rds_snapshot.go (hunk 8 above)
Comment (alexkayabula) [reply aws/rds_snapshot.go]: Fixed.

File: aws/rds_snapshot.go (hunk 10 above)
Comment (alexkayabula) [reply aws/rds_snapshot.go]: Fixed.

File: aws/rds_cluster_snapshots_test.go (hunk 6 above)
Comment (alexkayabula) [reply aws/rds_cluster_snapshots_test.go:128]: I am not sure. Is there a way to determine how long this test will run ?

File: aws/rds_cluster_snapshots.go [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+				if shouldIncludeClusterSnapshotByName(*snapshot.DBClusterSnapshotIdentifier, configObj.RDSSnapshots.IncludeRule.NamesRE, configObj.RDSSnapshots.ExcludeRule.NamesRE) {
+					if len(tagsResult.TagList) > 0 {
+						for _, tag := range tagsResult.TagList {
+							if shouldIncludeClusterSnapshotByTag(*tag.Key, configObj.RDSSnapshots.IncludeRule.TagNamesRE, configObj.RDSSnapshots.ExcludeRule.TagNamesRE) {
+								snapshots = append(snapshots, snapshot.DBClusterSnapshotIdentifier)
+							}
+						}
+					} else {
+						snapshots = append(snapshots, snapshot.DBClusterSnapshotIdentifier)
+					}
```
[the remark this answers is not in the document]
Comment (alexkayabula) [reply aws/rds_cluster_snapshots.go]: Fixed

File: aws/rds_snapshot_types.go (hunk 15 above)
Comment (alexkayabula) [reply aws/rds_snapshot_types.go]: Fixed

File: aws/rds_snapshot_test.go (hunk 13 above)
Comment (alexkayabula) [reply aws/rds_snapshot_test.go]: Fixed.

**Comment by brikis98** [comment]
@alexkayabula Unfortunately, that is a private Gruntwork account, and we only give employees access to it. AWS does offer a free tier, so if you spin up a new account, and stick to the smallest RDS instance types, or even serverless, I don't think it should cost anything...

**Comment by ina-stoyanova** [comment]
Giving this one a bump. It might be easiest to close this PR & revive it once we're in a good place to pick this issue up #120.

**Comment by ellisonc** [comment]
Closing this PR as it hasn't had activity in a while and will likely be easier to re-implement. Feel free to re-open at a later date.

**Pull request closed without merging**